sale_coupon_product_management/models/sale_coupon_program.py

```diff
@@ -0,0 +1,122 @@
+        readonly=False,
+        store=True,
+    )
+
+    @api.onchange("related_product_categ_id")
+    def _onchange_related_product_categ_id(self):
+        categ = self.related_product_categ_id
+        if categ._predicate_product_categ_with_opts():
+            # Updating only sale.coupon.program values here.
+            self.update(categ.program_option_ids.get_program_values()[SELF])
+
+    @api.onchange("promo_applicability")
+    def _onchange_promo_applicability(self):
+        if (
+            self.program_type == "promotion_program"
+            and self.promo_applicability == "on_next_order"
+        ):
+            default_categ = self.env["product.category"].search(
+                [("default_promotion_next_order_category", "=", True)], limit=1
+            )
+            if default_categ:
+                self.related_product_categ_id = default_categ.id
+
+    @api.constrains("related_product_categ_id", "reward_type", "discount_type")
+    def _check_program_options(self):
+        for rec in self:
+            categ = rec.related_product_categ_id
+            if categ._predicate_product_categ_with_opts():
+                categ.program_option_ids.validate_program(
+                    rec, excluded_paths=[DISCOUNT_PRODUCT_FNAME]
+                )
+
+    def _get_reward_line_product_extra_create_vals(self, vals):
+        self.ensure_one()
+        extra_vals = {}
+        for fname, val in vals.items():
+            # Only care about related product truthy vals.
+            if fname.startswith(PRODUCT_FNAME_PREFIX) and val:
+                # Remove prefix.
+                product_fname = fname[len(PRODUCT_FNAME_PREFIX) :]
+                extra_vals[product_fname] = val
+        return extra_vals
+
+    @api.model
+    def create(self, vals):
+        """Extend to initially update related product values.
+
+        Related fields on product won't be updated by default, because
+        product is created after related program creation.
+        """
+
+        def prepare_forced_product_vals():
+            forced_product_vals = {"name": vals["name"]}
+            categ = self.env["product.category"].browse(
+                vals.get("related_product_categ_id")
+            )
+            if categ._predicate_product_categ_with_opts():
+                options = categ.program_option_ids
+                extra_product_vals = options.get_program_values()[
+                    DISCOUNT_PRODUCT_FNAME
+                ]
+                forced_product_vals.update(extra_product_vals)
+            return forced_product_vals
+
+        self = self.with_context(forced_product_vals=prepare_forced_product_vals())
+        program = super(SaleCouponProgram, self).create(vals)
+        product_extra_vals = program._get_reward_line_product_extra_create_vals(vals)
+        if product_extra_vals:
+            program.discount_line_product_id.write(product_extra_vals)
+        return program
+
+    def write(self, vals):
```

Comment:
I am a little bit confused why we need such complexity to just update the product? 
we can make discount_line_product_id visible on sale.coupon.program and sale.coupon so customer can open a view and edit fields properly
cc @epanisset

Reply:
That field is already visible. And user can update it normally using standard functionality.

But requirement was to make it easier to update/manage program and its related product. Also program options allow to automate some things, so you only need to set category and then don't need to check if fields in question are set correctly.

Reply:
Yes, to avoid any mistake or missing information on their side + Because of the volume that would need to be handle, better to make it automatic :)